### PR TITLE
Reorganization of Color

### PR DESCRIPTION
**Pull request by z3dev**
This pull request adds the color functions as a seperate module. The goals are:
- Standardize API; names, parameter passing, error checks, etc.
- Full test coverage

color
- [X] code complete
- [X] tests complete, 81.25% coverage (due to error checks)
- [X] reveiw complete

Overall test coverage 83.87%.

### CODE REVIEW

**Comment by z3dev** [comment]
This could be the last reorganization.

@kaosat-dev I believe this is the correct location of the color module. Please review.

File: src/color/color.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+const colorNameToRgb = require('./colorNameToRgb')
+
+/**
+ * Apply the given color to the given objects.
+ * @param {Object} color - either an array or a hex string of color values
+ * @param {Object|Array} objects - the objects of which to color
+ * @returns {Object|Array} the same objects with an additional attribute 'color'
+ *
+ * @example
+ * let redSphere = color([1,0,0,1], sphere())
```
Comment (kaosat-dev) [inline src/color/color.js]: so the alpha for rgba is merged into the color array right ?
the only problem with that is for 'named/css' colors: they are rgb data, and without a seperate alpha parameter, we need to account for that somehow (hence some of the weirdness of the previous color op)

Comment (kaosat-dev) [reply src/color/color.js]: one possible solution, is to not do any 'magic' in here:
if you want a specific color, you (as a user) should create your colors using hexToRgba or colorNameToRgb etc
it has one extra perk : less code and conditionals here, and it means each functions does one thing only but one thing well :)

File: src/color/hexToRgb.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
+/**
+ * Converts a HTML5 color value (string) to RGB values.
+ * Conversion formula:
+ * - split the string; "#RRGGBB" into RGB components
+ * - convert the HEX value into RGB values
+ * @see the color input type of HTML5 forms
+ */
+const hexToRgb = (value) => {
```
Comment (kaosat-dev) [inline src/color/hexToRgb.js]: I think this needs to work for rgbA hex values, (even more so if we remove the alpha parameter from the color() op), 
we already have such a thing here https://github.com/jscad/OpenJSCAD.org/blob/V2/packages/utils/img/src/colorConversions.js
This is one of the things I mentionned needing cleanup after we merge Openjscad.org & csg, so we can deal with it then ?

File: src/color/hslToRgb.js [hunk 3]
```diff
@@ -1,21 +1,21 @@
 const hueToColorComponent = require('./hueToColorComponent')
 
 /**
- * Converts an HSL color value to RGB. Conversion formula
- * adapted from http://en.wikipedia.org/wiki/HSL_color_space.
+ * Converts an HSL color value to RGB.
  * Assumes h, s, and l are contained in the set [0, 1] and
  * returns r, g, and b in the set [0, 1].
  *
+ * @see http://en.wikipedia.org/wiki/HSL_color_space.
```
Comment (kaosat-dev) [inline src/color/hslToRgb.js:6]: nice !

File: src/color/hslToRgb.js [hunk 4]
```diff
@@ -1,21 +1,21 @@
...
  * @param   Number  h       The hue
  * @param   Number  s       The saturation
  * @param   Number  l       The lightness
  * @return  Array           The RGB representation
  */
 const hslToRgb = (h, s, l) => {
```
Comment (kaosat-dev) [inline src/color/hslToRgb.js]: same remark as above, perhaps h, s, l , a to get a toRgba ?

File: src/color/hsvToRgb.js [hunk 5]
```diff
@@ -1,19 +1,19 @@
...
- * @param   Number  s       The saturation
- * @param   Number  v       The value
- * @return  Array           The RGB representation
+ * @see http://en.wikipedia.org/wiki/HSV_color_space.
+ * @param Number h - hue
+ * @param Number s - saturation
+ * @param Number v - value
+ * @return Array the RGB representation
  */
 const hsvToRgb = (h, s, v) => {
```
Comment (kaosat-dev) [inline src/color/hsvToRgb.js]: same

File: src/color/index.js [hunk 6]
```diff
@@ -0,0 +1,12 @@
+module.exports = {
```
Comment (kaosat-dev) [inline src/color/index.js:1]: nice & clean !

File: src/color/rgbToHex.js [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+ * @param  {Float} g - green value
+ * @param  {Float} b - blue value
+ */
+const rgbToHex = (r, g, b) => {
+  if (Array.isArray(r)) {
+    b = r[2]
+    g = r[1]
+    r = r[0]
+  }
+  let s = '#' +
```
Comment (kaosat-dev) [inline src/color/rgbToHex.js]: es6 strings please :)

File: src/color/rgbToHsl.js [hunk 8]
```diff
@@ -1,17 +1,17 @@
...
- * @return  Array           The HSL representation
+ * @see http://axonflux.com/handy-rgb-to-hsl-and-rgb-to-hsv-color-model-c
+ * @param Number r - red color value
+ * @param Number g - green color value
+ * @param Number b - blue color value
+ * @return Array the HSL representation
  */
 const rgbToHsl = (r, g, b) => {
-  if (r.length) {
+  if (Array.isArray(r)) {
```
Comment (kaosat-dev) [inline src/color/rgbToHsl.js]: way better !

File: src/color/rgbToHsv.js [hunk 9]
```diff
@@ -1,20 +1,19 @@
...
- * adapted from http://en.wikipedia.org/wiki/HSV_color_space.
+ * Converts an RGB color value to HSV.
  * Assumes r, g, and b are contained in the set [0, 1] and
  * returns h, s, and v in the set [0, 1].
  *
- * @param   Number  r       The red color value
- * @param   Number  g       The green color value
- * @param   Number  b       The blue color value
- * @return  Array           The HSV representation
+ * @see http://en.wikipedia.org/wiki/HSV_color_space.
```
Comment (kaosat-dev) [inline src/color/rgbToHsv.js:4]: nice

**Review by kaosat-dev — CHANGES_REQUESTED** [review]
clean and to the point @z3dev ! Only point to discuss is the alpha value related issues.

File: src/color/color.js (hunk 1 above)
Comment (z3dev) [reply src/color/color.js]: I was thinking very similar. 

I think the color() function can just set color. And like you mentioned, the other functions can be used to translate between specifications or names.

But, I think most people will not provide the ALPHA.

We can have the color() function provide an ALPHA if not given (which the current code does)

OR

Like you said, let the viewer interpret colors if there’s no ALPHA.

Probably, we should do both.

File: src/color/rgbToHex.js (hunk 7 above)
Comment (z3dev) [reply src/color/rgbToHex.js]: I saw it... but the conversation to string is a wild one.

**Comment by z3dev** [comment]
I will add some support for HEX ALPHA, RGBA and HSLA, as per web standards.

https://www.w3schools.com/cssref/css_colors_legal.asp

@kaosat-dev how about removing support for passing arrays? I don’t like the code that suppprts arrays, and actually had to fix most of the logic.

File: src/color/color.js (hunk 1 above)
Comment (kaosat-dev) [reply src/color/color.js]: agreed, default to 1, but give the option to provide it (do  not give the option to set it in the color() function, but in the 'creator'/ converter function) :)

**Comment by kaosat-dev** [comment]
@z3dev do you mean arrays passed to the various conversion functions ? I think that makes sense, users can use funct(...array) to have the same effect right ?
hmm then again, one of the foundations of V2 was the use of arrays (which is very practical in a ton of cases).
or switch to JUST arrays ?

File: src/color/hexToRgb.js (hunk 2 above)
Comment (z3dev) [reply src/color/hexToRgb.js]: Changed to accept an string notiation with RGB or RGBA hex values.

File: src/color/hslToRgb.js (hunk 4 above)
Comment (z3dev) [reply src/color/hslToRgb.js]: Changed to accept an array of HSL or HSLA values.

File: src/color/hsvToRgb.js (hunk 5 above)
Comment (z3dev) [reply src/color/hsvToRgb.js]: Changed to accept an array of HSV or HSVA values.

**Comment by z3dev** [comment]
>  or switch to JUST arrays ?

All conversion functions have been changed to accept an array of RGB/HSL/HSV values, only. I think this looks far better.

@kaosat-dev please review again

**Comment by z3dev** [comment]
I created a new issue #187 to review these changes as part of the cleanup phase.

@kaosat-dev please merge if satisfied.

**Comment by kaosat-dev** [comment]
well done @z3dev !

**Pull request merged**